# Wonderful world of Structural Variant Calling! 

***
Lead: Fritz J Sedlazeck

Human Genome Sequencing Center
Baylor College of Medicine

Computer Science Department
Rice Univeristy 
***

## Goals of this module
The goal of this module is to get you familiarized with variant identification (especially Structural Variations) across assembly, short read based mapping and long read based mapping. 
For Structural Variations (SVs) detection we will be usinig mutliple methods (Assemblytics, Manta, Sniffles) and then later compare them across to obtain more insights about advantages and disadvantes across the differnet approaches.

### The main steps in this Module are:
1. Assembly based SV detection (using Assemblytics).
2. Short read mapping based SV detection (using ([Manta](https://github.com/Illumina/manta))). 
3. Long read based mapping based SV detection (using ([Sniffles](https://github.com/fritzsedlazeck/Sniffles))). 
4. SV comparison (using ([Sniffles](https://github.com/fritzsedlazeck/SURVIVOR)))

## Organism
We will utilize data from Cryptosporidium for our exercise. Cryptosporidium is a microscopic parasite that causes the diarrheal disease cryptosporidiosis. Both the parasite and the disease are commonly known as “Crypto.” There are many species of Cryptosporidium that infect animals, some of which also infect humans.
The data is from a recent publication in Gigascience: [Fully resolved assembly of Cryptosporidium parvum](https://doi.org/10.1093/gigascience/giac010). Here we produced short read Illumina data together wiht Oxford Nanopore long read data to obtain a fully assembled (T2T) assembly and thus improve the reference overall. 
We will study this sample of Cryptosporidium compared to the previously available reference sequence (GCF_000165345.1.). 

